lib/Sema/TypeCheckAccess.cpp

```diff
@@ -1421,43 +1426,104 @@ class ImplementationOnlyImportChecker
       const_cast<TypeRepr *>(typeRepr)->walk(TypeReprIdentFinder(
           [&](const ComponentIdentTypeRepr *component) {
         ModuleDecl *M = component->getBoundDecl()->getModuleContext();
-        if (!SF->isImportedImplementationOnly(M))
+        if (!SF.isImportedImplementationOnly(M))
           return true;
 
-        diagnose(component->getBoundDecl(), component);
+        diagnoseType(component->getBoundDecl(), component);
         foundAnyIssues = true;
         // We still continue even in the diagnostic case to report multiple
         // violations.
         return true;
       }));
     }
 
-    // FIXME: Are there places where we can be sure the TypeRepr fully accounts
-    // for the type, and therefore we could skip this extra work?
-    if (!foundAnyIssues && type) {
-      type.walk(SimpleTypeDeclFinder([&](const TypeDecl *typeDecl) {
+    // Note that if we have a type, we can't skip checking it even if the
```

Comment:
In theory we could also skip walking the type if our source file doesn't import anything as implementation-only, but that might be too much conditionalizing for little gain.

Reply:
Yeah, John had his part skipping. Preliminary "test compiler performance" didn't show any significant difference, but it's probably worth following up here to make sure that really is the case.